ksqldb-engine/src/main/java/io/confluent/ksql/execution/ExecutionPlanBuilder.java

```diff
@@ -87,4 +97,142 @@ public ExecutionPlan buildPhysicalPlan(
         resultStream.getSourceStep()
     );
   }
+
+
+  private void optimizeForSelfJoin(final PlanNode currentNode) {
+
+    if (currentNode instanceof JoinNode) {
+      final JoinNode joinNode = (JoinNode) currentNode;
+      if (! matchSubGraphPattern1(joinNode)) {
+        if (!matchSubGraphPattern2(joinNode)) {
+          matchSubGraphPattern3(joinNode);
```

Comment:
This pattern was non-obvious to me at first.  I wonder if there are any ideas for increasing clarity.

Also, as a nit, there's a space between the `!` and `matchSubGraphPattern1`.

Reply:
Does the comment above (https://github.com/confluentinc/ksql/pull/9490/files#r970827259) answer your questions @jnh5y ?